top_ten <- head(mergers, n = 10)

library(dada2); packageVersion("dada2")

setwd('~/LeBoldus/local_git/poplar_microbiome_REEU')

seqtab_ten <- makeSequenceTable(top_ten)

seqtab.nochim_ten <- removeBimeraDenovo(seqtab_ten, method="consensus", multithread=TRUE, verbose=TRUE)

dim(seqtab_ten)
dim(seqtab.nochim_ten)

# Define UNITE reference file path
unite.ref <- "UNITE/sh_general_release_29.11.2022/sh_general_release_dynamic_29.11.2022.fasta"

#Assign Taxonomy
taxa.ten <- assignTaxonomy(seqtab.nochim_ten, unite.ref, multithread = TRUE, tryRC = TRUE)

seqtab.nochim_ten <- head(seqtab.nochim, n = 10)
dim(seqtab.nochim_ten)

rm(dadaFs)
rm(dadaRs)
rm(derepFs)
rm(derepRs)
rm(err_F_plot)
rm(err_R_plot)
rm(errF)
rm(errR)
rm(fout)
rm(fout_df)
rm(mergers)
rm(nout)
rm(Q20_filtered_out)
rm(seqtab)
rm(seqtab.nochim.df.20)
rm(track)
rm(track_df)
rm(track_df_melt)
rm(track_fout)
rm(track_nout)
rm(track_plot)
rm(cut_path)
rm(cutadapt)
rm(data_path)
rm(filtered_path)
rm(filtered_path_Q20)
rm(filtFs)
rm(filtRs)
rm(fnFs_cut)
rm(fnFs_filtN)
rm(fnFs)
rm(fnRs)
rm(fnRs_cut)
rm(fnRs_filtN)
rm(FWD)
rm(FWD.orients)
rm(FWD.RC)
rm(i)
rm(ITS3_KYO1)
rm(ITS4)
rm(n_path)
rm(outputs_path)
rm(post_filter_fnFs)
rm(post_filter_fnRs)
rm(post_filter_sample_names)
rm(Q20_path)
rm(Q28_path)
rm(R1.flags)
rm(R2.flags)
rm(REV)
rm(REV.orients)
rm(REV.RC)
rm(sample_names)
rm(sample_numbers)
rm(x)
rm(Q20_track_head)
rm(allOrients)
rm(getN)
rm(primerHits)

dim(seqtab.nochim)
